### PR TITLE
[TorchMLIR]add hardtanh op

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp [hunk 1]
```diff
@@ -44,8 +43,7 @@ std::unordered_set<std::string> GetTorchMlirWhiteList() {
       "aten::mul",      "aten::eq",          "aten::lt",
       "aten::ne",       "aten::__and__",     "aten::floor_divide",
       "prim::Constant", "aten::expand_as",   "aten::repeat",
-      "aten::expand",
-  };
+      "aten::expand",   "aten::hardtanh"};
```
Comment (wyzero) [inline pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp:46]: Should we sort this op list by name (e.g. one line on op, disable clang-format here)？

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 2]
```diff
@@ -796,6 +796,37 @@ LogicalResult ConvertAtenOp<AtenGeluOp>::matchAndRewrite(
...
+    return rewriter.notifyMatchFailure(
+        op, "non-const size parameter unsupported");
+  auto opType = getTypeConverter()
+                    ->convertType(op.getType())
+                    .template cast<RankedTensorType>();
+  if (!opType) {
+    return op.emitError("should be RankedTensorType");
+  }
+  Type resultElementType;
+  if (op.dtype().getType().isa<Torch::NoneType>()) {
```
Comment (fortianyou) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:816]: Why should `dtype` force to none?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 3]
```diff
@@ -796,6 +796,37 @@ LogicalResult ConvertAtenOp<AtenGeluOp>::matchAndRewrite(
...
+    resultElementType = opType.getElementType();
+  } else {
+    return op.emitError("dtype should be none");
+  }
+
+  if (!resultElementType.isa<mlir::FloatType>()) {
```
Comment (fortianyou) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:822]: Add i64, i32, float64, bool support?

File: pytorch_blade/tests/disc/ops/test_activation.py [hunk 4]
```diff
@@ -55,12 +55,13 @@ def test_gelu_static_shape(self):
     def test_gelu_dynamic_shape(self):
         self._test_activation(torch.nn.GELU(), torch.nn.functional.gelu, [([-1, -1, -1, -1], torch.float)])
 
-
-    @skipIfEnableTorchMlir()
-    #TODO(yancey1989): dependence binary op
-    def test_hardtanh(self):
+    def test_hardtanh_static_shape(self):
```
Comment (fortianyou) [inline pytorch_blade/tests/disc/ops/test_activation.py:58]: Add empty format test?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp (hunk 2 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:816]: image

File: pytorch_blade/tests/disc/ops/test_activation.py (hunk 4 above)
Comment (fortianyou) [reply pytorch_blade/tests/disc/ops/test_activation.py:58]: image

File: pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp (hunk 1 above)
Comment (Yancey1989) [reply pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp:46]: agree with you, will do that.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp (hunk 2 above)
Comment (Yancey1989) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:816]: only supports None in this PR, will supports more dtype and add an unit test for the empty tensor.

File: pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp (hunk 1 above)
Comment (Yancey1989) [reply pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp]: this enhancement has implemented in another PR.

File: pytorch_blade/tests/disc/ops/test_activation.py (hunk 4 above)
Comment (Yancey1989) [reply pytorch_blade/tests/disc/ops/test_activation.py:58]: I removed the empty tensor lowering, the empty tensor is lowered from the official decompose pass, this action is not as disc expected, so I added a `DiscDecompose` pass in front of the official decompose pass.

File: pytorch_blade/src/torch-mlir/lib/Dialect/TorchConversion/Transforms/DiscDecomposeComplexOps.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  BaseTensorType inputType = input.getType().cast<BaseTensorType>();
+
+  auto sizeListType =
+      Torch::ListType::get(Torch::IntType::get(op.getContext()));
+  Value sizeList =
+      rewriter.create<AtenSizeOp>(op.getLoc(), sizeListType, input);
+
+  SmallVector<int64_t> empty;
+  Type tensorType = inputType.getWithSizesAndDtype(
+      llvm::makeArrayRef(empty), rewriter.getF32Type());
```
Comment (fortianyou) [inline pytorch_blade/src/torch-mlir/lib/Dialect/TorchConversion/Transforms/DiscDecomposeComplexOps.cpp:63]: Nit: inputType.getElementType(), you can fix next PR.

**Review by fortianyou — APPROVED** [review]
LGTM
